compiler/rustc_mir/src/shim.rs

```diff
@@ -145,7 +145,18 @@ fn build_drop_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Option<Ty<'tcx>>)
     };
     let sig = tcx.fn_sig(def_id).subst(tcx, substs);
     let sig = tcx.erase_late_bound_regions(sig);
-    let span = tcx.def_span(def_id);
+
+    // Use the definition of the type for which this is the drop glue as span when possible instead
+    // of the span of `core::mem::drop_in_place`.
+    let span = if let Some(ty) = ty {
+        match ty.kind() {
+            ty::Adt(adt_def, _) => tcx.def_span(adt_def.did),
+            ty::Foreign(def_id) => tcx.def_span(*def_id),
+            _ => tcx.def_span(def_id),
+        }
```

Comment:
AFAICT this would fail and fall-back to the original behaviour for types such as `[Adt; N]`, tuples of `Adt`s, etc.

Previously the behaviour was at least universally consistent regardless of type. So to me it seems like making a change such as this could do more harm than good.

Reply:
I dropped this change.